### ISSUE TITLE
com.vdurmont.semver4j.SemverException: Invalid version (no major version): "^1.2.0"

### ISSUE THREAD

**Issue by AshleyEke** [dependency-check/DependencyCheck#3956]
Since an upgrade to 6.5.2 we are getting an error when running dependency check. [Log Here](https://gist.githubusercontent.com/AshleyEke/7a0a43624018b5f5b5b0ce630c7fcd4c/raw/6768b73aa6b7f9095a8be2d6f2e81fc6cd7ba1e9/dependency-check-semver-error.txt)

I suspect [this](https://github.com/jeremylong/DependencyCheck/pull/3909/files) change is the cause. Seems like semver isn't handling the ^ symbol correctly.

**Comment by jeremylong** [comment]
Do you have a sample project? I'm betting this is due to the Node.js Package Analyzer being enabled -  we have plans to deprecate this analyzer and fold some of its capabilities into the Node Audit Analyzer.

**Comment by aikebah** [comment]
I would, however, not expect the package-lock.json to have a semver range version. Always thought that package-lock.json was meant to keep development state at an explicit version and only package.json was supposed to have ranged versions?

https://docs.npmjs.com/cli/v6/configuring-npm/package-locks

**Comment by mcwarman** [comment]
The error can be re-created by:

`package.json`

```json
{
  "name": "my-awesome-package",
  "version": "1.0.0",
  "dependencies": {
    "minimist": "^1.2.5",
    "sharkdown": "^0.1.1"
  }
}
```

`package-lock.json`

```json
{
  "name": "my-awesome-package",
  "version": "1.0.0",
  "lockfileVersion": 1,
  "requires": true,
  "dependencies": {
    "ansicolors": {
      "version": "0.2.1",
      "resolved": "https://registry.npmjs.org/ansicolors/-/ansicolors-0.2.1.tgz",
      "integrity": "sha1-vgiVmQl7dKXJxKhKDNvNtivYeu8="
    },
    "cardinal": {
      "version": "0.4.4",
      "resolved": "https://registry.npmjs.org/cardinal/-/cardinal-0.4.4.tgz",
      "integrity": "sha1-ylu2iltRG5D+k7ms6km97lwyv+I=",
      "requires": {
        "ansicolors": "~0.2.1",
        "redeyed": "~0.4.0"
      }
    },
    "esprima": {
      "version": "1.0.4",
      "resolved": "https://registry.npmjs.org/esprima/-/esprima-1.0.4.tgz",
      "integrity": "sha1-n1V+CPw7TSbs6d00+Pv0drYlha0="
    },
    "minimist": {
      "version": "1.2.5",
      "resolved": "https://registry.npmjs.org/minimist/-/minimist-1.2.5.tgz",
      "integrity": "sha512-FM9nNUYrRBAELZQT3xeZQ7fmMOBg6nWNmJKTcgsJeaLstP/UODVpGsr5OhXhhXg6f+qtJ8uiZ+PUxkDWcgIXLw=="
    },
    "redeyed": {
      "version": "0.4.4",
      "resolved": "https://registry.npmjs.org/redeyed/-/redeyed-0.4.4.tgz",
      "integrity": "sha1-N+mQpvKyGyoRwuakj9QTVpjLqX8=",
      "requires": {
        "esprima": "~1.0.4"
      }
    },
    "sharkdown": {
      "version": "0.1.1",
      "resolved": "https://registry.npmjs.org/sharkdown/-/sharkdown-0.1.1.tgz",
      "integrity": "sha512-exwooSpmo5s45lrexgz6Q0rFQM574wYIX3iDZ7RLLqOb7IAoQZu9nxlZODU972g19sR69OIpKP2cpHTzU+PHIg==",
      "requires": {
        "cardinal": "~0.4.2",
        "minimist": "0.0.5",
        "split": "~0.2.10"
      },
      "dependencies": {
        "minimist": {
          "version": "0.0.5",
          "resolved": "https://registry.npmjs.org/minimist/-/minimist-0.0.5.tgz",
          "integrity": "sha1-16oye87PUY+RBqxrjwA/o7zqhWY="
        }
      }
    },
    "split": {
      "version": "0.2.10",
      "resolved": "https://registry.npmjs.org/split/-/split-0.2.10.tgz",
      "integrity": "sha1-Zwl8YB1pfOE2j0GPBs0gHPBSGlc=",
      "requires": {
        "through": "2"
      }
    },
    "through": {
      "version": "2.3.8",
      "resolved": "https://registry.npmjs.org/through/-/through-2.3.8.tgz",
      "integrity": "sha1-DdTJ/6q8NXlgsbckEV1+Doai4fU="
    }
  }
}
```

```shell
dependency-check --scan ./package-lock.json
```

```text
[WARN] An unexpected error occurred during analysis of '.../tmp/package-lock.json' (Node Audit Analyzer): Invalid version (no major version): "^1.2.5"
[ERROR]
com.vdurmont.semver4j.SemverException: Invalid version (no major version): "^1.2.5"
	at com.vdurmont.semver4j.Semver.<init>(Semver.java:61)
	at org.owasp.dependencycheck.analyzer.AbstractNpmAnalyzer.determineVersionFromMap(AbstractNpmAnalyzer.java:530)
	at org.owasp.dependencycheck.analyzer.AbstractNpmAnalyzer.processResults(AbstractNpmAnalyzer.java:471)
	at org.owasp.dependencycheck.analyzer.NodeAuditAnalyzer.analyzeDependency(NodeAuditAnalyzer.java:151)
	at org.owasp.dependencycheck.analyzer.AbstractAnalyzer.analyze(AbstractAnalyzer.java:131)
	at org.owasp.dependencycheck.AnalysisTask.call(AnalysisTask.java:88)
	at org.owasp.dependencycheck.AnalysisTask.call(AnalysisTask.java:37)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

The error is confusing as there is no mention of version `^1.2.5` in `package-lock.json`, which i'm guessing is being picked up from the `package.json`

Running it with `--disableNodeAudit` causes it not to error.

**Comment by mebbage** [comment]
**Issue is not limited to semver ranges**

We've also hit the same error, after going from 6.2.2 to 6.5.2

     [ERROR] 
      com.vdurmont.semver4j.SemverException: Invalid version (no major version): "6.0.1"
	  at com.vdurmont.semver4j.Semver.<init>(Semver.java:61)
	  at org.owasp.dependencycheck.analyzer.AbstractNpmAnalyzer.determineVersionFromMap(AbstractNpmAnalyzer.java:530)
	  at org.owasp.dependencycheck.analyzer.AbstractNpmAnalyzer.processResults(AbstractNpmAnalyzer.java:471)
	  at org.owasp.dependencycheck.analyzer.YarnAuditAnalyzer.analyzeDependency(YarnAuditAnalyzer.java:108)
	  at org.owasp.dependencycheck.analyzer.AbstractAnalyzer.analyze(AbstractAnalyzer.java:131)
	  at org.owasp.dependencycheck.AnalysisTask.call(AnalysisTask.java:88)
	  at org.owasp.dependencycheck.AnalysisTask.call(AnalysisTask.java:37)
	  at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	  at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	  at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	  at java.base/java.lang.Thread.run(Thread.java:832)
    [INFO] Finished Yarn Audit Analyzer (2 seconds)

In 6.2.2 the Yarn Audit Analyzer worked as expected - the exception we now get is a bit cryptic - as the semver appears valid (there is a major version) and the log doesn't indicate which library (in our `yaml.lock`) has the issue?

**Comment by mebbage** [comment]
Your `package.json` specifies the required version of minimist to be `^1.2.5`.

Your `package-lock.json` contains the resolved version `1.2.5`:

>     "minimist": {
>       "version": "1.2.5",
>       "resolved": "https://registry.npmjs.org/minimist/-/minimist-1.2.5.tgz",
>       "integrity": "sha512-FM9nNUYrRBAELZQT3xeZQ7fmMOBg6nWNmJKTcgsJeaLstP/UODVpGsr5OhXhhXg6f+qtJ8uiZ+PUxkDWcgIXLw=="
>     },
> The error is confusing as there is no mention of version `^1.2.5` in `package-lock.json`, which i'm guessing is being picked up from the `package.json`

Your `package-lock.json` will not generally include the tilde and caret notations (specifying tolerance around updated minor and patch versions) except in the `requires` object - not for the resolved dependencies, as seen in your example.

A `package-lock.json` using `^` or `~` rather than version pinning would kinda defeat the point of its existence. But you're right, the error output from Node Audit Analyzer implies the referenced version is in your lock file - which it clearly isn't.

This wasn't obvious in my example, because the referenced version (6.0.1) does appear in my `yarn.lock` (coincidentally, for several unrelated libraries).
